### ISSUE TITLE
Cats Order needs an `orElseBy` combinator

### ISSUE THREAD

**Issue by benhutchison** [typelevel/cats#4621]
Leaving aside the question of whether `cats.Order` existing (on top of `scala.math.Ordering` on top of `java.util.Comparator`) was wise decision 🙄 , at least it ought to be greaterThanOrEqual to its predecessor.

But alas, it lacks something valuable that Ordering has; a convenient syntax to construct a n-level hierarchical ordering for a record, by delegating to the orderings of several record fields. For example, if we have a `case class Person(name: String, age: Int)` we might wish to order by `age` but if ages are equal, use `name` as a discriminator. This is super common IME. Eg SQL has built in syntax for it `order by age, name`.

There are at least two attempts at addressing this in cats.Order, but alas both are less convenient than the (pre-existing) combinator in `scala.math.Ordering`:

`def orElseBy[S](f: T => S)(implicit ord: Ordering[S]): Ordering[T]`
  
 Hence we find code like this[ in the wild](https://gist.github.com/REDNBLACK/014c7c56cd78b2a92ea4bf507621fad3#file-generator-scala-L94):
  ```
        given Order[Single] = Order.fromOrdering(
        Ordering
          .by[Single, VPNTier](_.tier)
          .orElseBy(it => (it.lang1, it.lang2))
          .orElseBy(_.num)
          .orElseBy(_.postfix)
      )
  ```
  
  The attempts in cats.Order are:
  - [code](https://github.com/typelevel/cats/blob/main/kernel/src/main/scala/cats/kernel/Order.scala#L174). `Order.whenEqual[Person](Order.by(_.age), Order.by(_.name))`. Well it's already looking slightly awkward, but what if we wanted to add a 3rd discriminator? We'd need to nest it inside the second arg with another call to Order.whenEqual. Confusing right nesting.
  - [code](https://github.com/typelevel/cats/blob/main/kernel/src/main/scala/cats/kernel/Order.scala#L223). The `whenEqual` Monoid is beautiful and tantalizes us with the promise of `val o: Order[Person] = Order.by(_.age) |+| Order.by(_.name)`. But I could not get the types to infer properly even on Scala 3.5 RC1.
  
If it ain't broke, don't fix it. `orElseBy` does the job well and should be brought into Cats.

**Comment by johnynek** [comment]
Seems like a good PR.

I don't know all the motivations for Order but I think it was that Ordering I believe was contravariant and scala had (or maybe has?) some issues with very surprising implicit resolution on contravariant type classes.

**Comment by satorg** [comment]
I think it would be a great addition indeed.

However, I wonder why `orElse` and `orElseBy` were added to `Ordering`, but not `PartialOrdering`.
And whether we'd like to address it in `PartialOrder` instead.

**Comment by benhutchison** [comment]
>  I wonder why orElse and orElseBy were added to Ordering, but not PartialOrdering

What would be the semantics of `orElse` etc when two elements are not comparable on the primary PartialOrder? ie `tryCompare` returns `None`. presumably theres no delegation for consistency. The name `orElse` could lead people to think it tries another PO if the first is incomparable. 

That semantic does seem workable . Perhaps the slight ambiguity, or simply that PartialOrdering is less commonly used, is the reason.

**Comment by satorg** [comment]
That's right. So I think that in theory, `PartialOrder` could offer two different `orElse` like combinators, e.g. `orElseOnEqual` and `orElseOnUnrelated`. Then in `Order` the former one should be simply inherited without changes, while the latter could short-circuit in order to simply return a result of "this" `Order` instance.

Not sure if all that makes sense though, wdyt?

**Comment by benhutchison** [comment]
A [precedent](https://github.com/zio/zio-prelude/blob/series/2.x/core/shared/src/main/scala/zio/prelude/Ord.scala#L897) for Partial order `orElse` combinator

> That's right. So I think that in theory, PartialOrder could offer two different orElse like combinators, e.g. orElseOnEqual and orElseOnUnrelated

I'm reluctant to implement a `orElseOnUnrelated` combinator without a clear use case. 

> Then in Order the former one should be simply inherited without changes

In spirit, yes. But partial orders return a comparison wrapped in an Option, while total orders omit the Option.

Interestingly, ZIO nicely avoids an Option by using a 4-branch sealed trait (LT, Eq, GT, Incomparable), which nests a 3-branch trait used for total orders 🆒

**Comment by satorg** [comment]
>In spirit, yes. But partial orders return a comparison wrapped in an Option, while total orders omit the Option.

> Interestingly, ZIO nicely avoids an Option by using a 4-branch [sealed trait ](https://github.com/zio/zio-prelude/blob/series/2.x/core/shared/src/main/scala/zio/prelude/Ord.scala#L858)(LT, Eq, GT, Incomparable), which nests a 3-branch trait used for total orders 🆒

Actually, in Cats the `tryCompare` method (which returns `Option`) is complimentary. The primary one is `partialCompare`:
https://github.com/typelevel/cats/blob/dd891334505b35e010a12da254924811de838c85/kernel/src/main/scala/cats/kernel/PartialOrder.scala#L59-L80
So it still should be quite efficient.

**Comment by satorg** [comment]
> I'm reluctant to implement a orElseOnUnrelated combinator without a clear use case.

Agree. I'm just trying to poke around different possibilities and outcomes. For example, if a method can be added to a base trait, it is usually better to add it there from the beginning even if it is not needed right away. Because if we decided to add it to the base class later, then it would be more difficult to overcome inevitable binary compatibility issues.

But when it comes to `orElseOnUnrelated`, it does not seem to be the case anyway – looks like it can be added later if necessary without any consequences (unless I'm missing something).

**Comment by morgen-peschke** [comment]
> What would be the semantics of `orElse` etc when two elements are not comparable on the primary PartialOrder? ie `tryCompare` returns `None`. presumably theres no delegation for consistency. The name `orElse` could lead people to think it tries another PO if the first is incomparable.

A possible fix for this provides a nice parallel with the existing method: name it `whenEqual` and expose it on the instance rather than the companion object.

This seems to imply fallbacks in any case:
```scala
PartialOrdering
  .by[Single, VPNTier](_.tier)
  .orElseBy(it => (it.lang1, it.lang2))
  .orElseBy(_.num)
  .orElseBy(_.postfix)
```
This doesn't have that connotation (at least to my reading):
```scala
PartialOrdering
  .by[Single, VPNTier](_.tier)
  .whenEqual(it => (it.lang1, it.lang2))
  .whenEqual(_.num)
  .whenEqual(_.postfix)
```

**Comment by mattjliu** [comment]
Hi, I'd be interested in taking this on. @satorg do you still think an `orElseUnrelated` combinator is necessary, or should I just go ahead and implement `orElseOnEqual`?

I'm somewhat partial to @morgen-peschke's suggestion of naming it `whenEqual`, but `Order.whenEqual` is in the companion object, which would be inconsistent with having it as an instance method for `PartialOrder`.

Alternatively I can open a PR when I'm done and we can discuss it then, thanks!

**Comment by satorg** [comment]
@mattjliu , thank you for your interest in this issue!

> but Order.whenEqual is in the companion object, which would be inconsistent with having it as an instance method for PartialOrder.

`whenEqualThenBy` / `whenUnrelatedThenBy` maybe?

> do you still think an orElseUnrelated combinator is necessary, or should I just go ahead and implement orElseOnEqual?

Adding the unrelated-based combinator is not a big issue. A more interesting question: is it possible to implement `whenEqualThenBy` in `PartialOrder` in such a way that
- when it is called on a `PartialOrder` instance, then it should take `implicit B: PartialOrder[B]` and return `PartialOrder[A]`.
- But if it is called on `Order` instance, it should require `implicit B: Order[B]` and return `Order[A]` instead.

The Scala's `Ordering`/`PartialOrdering` seem to ignore that issue completely.
However, there are some clues in the Scala's collection implementations (see `Seq`/`SeqOps`/`Iterable`/`IterableOps`). I wonder if it makes sense to explore those approaches.

**Comment by mattjliu** [comment]
Hi @satorg, thanks for the reply!

For reference, [SeqOps](https://github.com/scala/scala/blob/3d08a437bf5c2493ebc44c658fd3edbd8b20c468/src/library/scala/collection/Seq.scala#L77) uses this approach:

```scala
/** Base trait for Seq operations
  *
  * @tparam A the element type of the collection
  * @tparam CC type constructor of the collection (e.g. `List`, `Set`). Operations returning a collection
  *             with a different type of element `B` (e.g. `map`) return a `CC[B]`.
  * @tparam C  type of the collection (e.g. `List[Int]`, `String`, `BitSet`). Operations returning a collection
  *             with the same type of element (e.g. `drop`, `filter`) return a `C`.
  ...
  */
trait SeqOps[+A, +CC[_], +C] extends Any with IterableOps[A, CC, C] { self =>
  ...
  def sorted[B >: A](implicit ord: Ordering[B]): C = { ... }
}
```

Then for any type `Foo[A]` they simply extend `SeqOps[A, Foo, Foo[A]]`, and `foo.sorted` will return `Foo[A]` (see [List](https://github.com/scala/scala/blob/3d08a437bf5c2493ebc44c658fd3edbd8b20c468/src/library/scala/collection/immutable/List.scala#L82-L83) and [Seq](https://github.com/scala/scala/blob/3d08a437bf5c2493ebc44c658fd3edbd8b20c468/src/library/scala/collection/Seq.scala#L27))

But it looks like immediately this won't work with `PartialOrder` and `Order`, since we would need something like:

```scala
trait PartialOrderOps[A, P[_]] {
  def whenEqualThenBy[B](f: A => B)(implicit B: P[B]): P[A]
}

trait PartialOrder[A] extends PartialOrderOps[A, PartialOrder]                                                                    
trait Order[A] extends PartialOrderOps[A, Order] with PartialOrder[A]
```

which would cause an illegal inheritance for `Order`.

I guess we can also just overload the method in `Order`, but that's not what we're looking for and probably error prone.

**Comment by satorg** [comment]
@mattjliu , how about something like this:
```scala
trait PartialOrder[@sp A] extends Any with Eq[A] with PartialOrderOps[A, PartialOrder] { self => ... }

private[cats] trait PartialOrderOps[A, +OO[a] <: PartialOrder[a]] extends Any { self: PartialOrder[A] =>
  def whenEqualThenBy[B, OOO[a] >: OO[a]](f: A => B)(implicit OO: OOO[B]): OO[A] = ???
}

trait Order[@sp A] extends Any with PartialOrder[A] with OrderOps[A, Order] { self => ... }

private[cats] trait OrderOps[A, +OO[a] <: Order[a]] extends Any with PartialOrderOps[A, OO] { self: Order[A] => }

```
(disclaimer: not really verified)

**Comment by mattjliu** [comment]
looks like this does in fact work for scala 2:

```scala
//> using scala 2.13

trait PartialOrderOps[A, +OO[a] <: PartialOrder[a]] extends Any { self: PartialOrder[A] =>
  def whenEqualThenBy[B, OOO[a] >: OO[a]](f: A => B)(implicit ev: OOO[B]): OO[A] = ???
}

trait PartialOrder[A] extends Any with PartialOrderOps[A, PartialOrder] {
  def partialCompare(x: A, y: A): Double
}

trait Order[A] extends Any with PartialOrder[A] with PartialOrderOps[A, Order] {
  def compare(x: A, y: A): Int
  override def partialCompare(x: A, y: A): Double = compare(x, y).toDouble
}

object Main extends App {
  implicit val intOrder: Order[Int] = (x, y) => x.compareTo(y)

  // Calling on Order with explicit OOO returns Order[A]
  def test1(): Order[Int] = (intOrder: Order[Int]).whenEqualThenBy[Int, Order](identity)

  // Calling on PartialOrder with explicit OOO returns PartialOrder[A]
  def test2(): PartialOrder[Int] = (intOrder: PartialOrder[Int]).whenEqualThenBy[Int, PartialOrder](identity)

  // Type inference
  def test3(): Order[Int] = (intOrder: Order[Int]).whenEqualThenBy(identity)

  // Chaining on Order stays Order[A]
  def test4(): Order[Int] = (intOrder: Order[Int]).whenEqualThenBy(identity).whenEqualThenBy(identity)

  // Chaining on PartialOrder stays PartialOrder[A]
  def test5(): PartialOrder[Int] = (intOrder: PartialOrder[Int]).whenEqualThenBy(identity).whenEqualThenBy(identity)

  println("All type checks passed!")
}
```

For scala 3 we need to add an upper bound to `OOO`, since the compiler will not do an implicit search as `OOO` is not specific enough (this also works for scala 2):

```scala
  def whenEqualThenBy[B, OOO[a] >: OO[a] <: PartialOrder[a]](f: A => B)(implicit ev: OOO[B]): OO[A] = ???
```

**Comment by satorg** [comment]
@mattjliu , thank you for exploring those options.

I have to confess that I enjoy this kind of type-level equilibrium a lot — which is why I suggested it in the first place.

But I also have to admit that the implementation becomes far too hairy. Perhaps it makes sense to go with the approach you mentioned before: method overloading, i.e.:
```scala
trait PartialOrder[@sp A] extends Any with Eq[A] {
  def whenEqualThenBy[B: PartialOrder](f: A => B): PartialOrder[A] = ???
}
trait Order[@sp A] extends Any with PartialOrder[A] {
  def whenEqualThenBy[B: Order](f: A => B): Order[A] = ???
}
```

I’m personally not a big fan of such approaches, because method resolution depends only on the implicit parameter. But on the other hand, it seems to make sense in this particular case: if `whenEqualThenBy` is called on `Order` and only `PartialOrder` is available for `B`, the call is absolutely legitimate, but the result has to be `PartialOrder`, not `Order`. The only drawback is that `whenEqualThenBy` will have to be implemented twice, I guess.

**Comment by mattjliu** [comment]
@satorg, this makes sense, but I am unable to make this compile for scala 2 or scala 3:

```scala
trait PartialOrder[A] {
  def partialCompare(x: A, y: A): Double
  def whenEqualThenBy[B: PartialOrder](f: A => B): PartialOrder[A] = { (x, y) =>
    val c = this.partialCompare(x, y)
    if (c == 0.0) implicitly[PartialOrder[B]].partialCompare(f(x), f(y)) else c
  }
}

trait Order[A] extends PartialOrder[A] {
  def compare(x: A, y: A): Int
  def partialCompare(x: A, y: A): Double = compare(x, y).toDouble
  def whenEqualThenBy[B: Order](f: A => B): Order[A] = { (x, y) =>
    val c = this.compare(x, y)
    if (c == 0) implicitly[Order[B]].compare(f(x), f(y)) else c
  }
}

object Main extends App {
  implicit val intOrder: Order[Int] = (x, y) => x.compareTo(y)
  implicit val stringPartialOrder: PartialOrder[String] = (x, y) => x.compareTo(y).toDouble

  // Should be PartialOrder[A] when only PartialOrder[B] available
  val test1: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString)

  // Same but chained
  val test2: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString).whenEqualThenBy[Int](identity)
}
```

```scala
Compiling project (Scala 2.13.18, JVM (17))
[error] ./test-overload.scala:25:34
[error] could not find implicit value for evidence parameter of type Order[String]
[error]   val test1: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString)
[error]                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error] ./test-overload.scala:29:34
[error] could not find implicit value for evidence parameter of type Order[String]
[error]   val test2: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString).whenEqualThenBy[Int](identity)
[error]                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error compiling project (Scala 2.13.18, JVM (17))
Compilation failed
```

```scala
Compiling project (Scala 3.8.3, JVM (17))
[error] ./test-overload.scala:25:78
[error] No given instance of type Order[String] was found for a context parameter of method whenEqualThenBy in trait Order
[error]   val test1: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString)
[error]                                                                              ^
[error] ./test-overload.scala:29:78
[error] No given instance of type Order[String] was found for a context parameter of method whenEqualThenBy in trait Order
[error]   val test2: PartialOrder[Int] = intOrder.whenEqualThenBy[String](_.toString).whenEqualThenBy[Int](identity)
[error]                                                                              ^
Error compiling project (Scala 3.8.3, JVM (17))
Compilation failed
```

Seems like the overload resolution resolves `Order` over `PartialOrder` regardless of implicit parameters. I guess it's always possible to make this compile by casting it explicitly:

```scala
val po: PartialOrder[Int] = (intOrder: PartialOrder[Int]).whenEqualThenBy[String](_.toString)
```

I'm kinda a noob when it comes to these advanced scala type system tricks, maybe there's a way to do this without relying on overload resolution?

**Comment by satorg** [comment]
@mattjliu, thank you for verifying that! I haven't verified it myself before writing the last message, sorry.

I think, in this case, the simplest and clearest approach would be to keep those methods fully separate. Consequently, the naming should become a bit more verbose, e.g.:
```scala
trait PartialOrder[@sp A] extends Any with Eq[A] {
  def whenEqualThenPartialOrderBy[B: PartialOrder](f: A => B): PartialOrder[A] = ???
}
trait Order[@sp A] extends Any with PartialOrder[A] {
  def whenEqualThenOrderBy[B: Order](f: A => B): Order[A] = ???
}
```
It might look too verbose, but given that it is supposed to be used for building typeclasses and not the ordering per se, the long names should be fine, I guess.

Feel free to suggest better names or even a better entire approach!
